presto-main/src/main/java/com/facebook/presto/operator/SpillableGroupedTopNBuilder.java

```diff
@@ -0,0 +1,304 @@
+        this.partitionTypes = partitionTypes;
+        this.partitionChannels = partitionChannels;
+        this.operatorContext = operatorContext;
+
+        initializeInMemoryGroupedTopNBuilder();
+
+        this.localUserMemoryContext = operatorContext.localUserMemoryContext();
+        this.localRevocableMemoryContext = operatorContext.localRevocableMemoryContext();
+
+        this.unspillMemoryLimit = unspillMemoryLimit;
+        this.producingOutput = false;
+    }
+
+    public Work<?> processPage(Page page)
+    {
+        checkState(hasPreviousSpillCompletedSuccessfully(), "Previous spill hasn't yet finished");
+
+        Work<?> result = inMemoryGroupedTopNBuilder.processPage(page);
+        updateMemoryReservations();
+        return result;
+    }
+
+    private boolean hasPreviousSpillCompletedSuccessfully()
+    {
+        if (spillInProgress.isDone()) {
+            // check for exception from previous spill for early failure
+            checkSpillSucceeded(spillInProgress);
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public WorkProcessor<Page> buildResult()
+    {
+        // spill could be in progress.
+        checkSpillSucceeded(spillInProgress);
+
+        producingOutput = true;
+        // Convert revocable memory to user memory as returned Iterator holds on to memory so we no longer can revoke.
+        if (localRevocableMemoryContext.getBytes() > 0) {
+            long currentRevocableBytes = localRevocableMemoryContext.getBytes();
+            localRevocableMemoryContext.setBytes(0);
+            if (!localUserMemoryContext.trySetBytes(localUserMemoryContext.getBytes() + currentRevocableBytes)) {
+                localRevocableMemoryContext.setBytes(currentRevocableBytes);
+                checkSpillSucceeded(spillToDisk());
+            }
+        }
+
+        if (!spiller.isPresent()) {
+            return inMemoryGroupedTopNBuilder.buildResult();
+        }
+
+        // Spill the remaining collected input (also clear the memory reservations)
+        // TODO: Possible Optimization here is to not spill the last remaining buffered input
+        // and instead do a memory+disk sort merge. SpillableHashAggregationBuilder does this
+        if (inMemoryGroupedTopNBuilder.getBufferedPages().size() > 0) {
+            checkSpillSucceeded(spillToDisk());
+            this.localRevocableMemoryContext.setBytes(0);
+        }
+
+        // Collect all spill streams to merge-sort
+        List<WorkProcessor<Page>> sortedPageStreams = ImmutableList.<WorkProcessor<Page>>builder()
+                .addAll(spiller.get().getSpills().stream()
+                        .map(WorkProcessor::fromIterator)
+                        .collect(toImmutableList()))
+                .build();
+
+        // Sort-Merge the rows and produce group-by-group output
+        return getFinalResult(sortedPageStreams);
+    }
+
+    @Override
+    public long getGroupIdsSortingSize()
```

Comment:
not quite understand the usage of this method. It is only used in `getEstimatedSizeInBytes`? So why not directly use `getEstimatedSizeInBytes`?